### ISSUE TITLE
panic: unexpected error during closedir "Bad file descriptor" for unix::fs::Dir

### ISSUE THREAD

**Issue by ConradIrwin** [rust-lang/rust#124105]
At [Zed](https://zed.dev) we see 2-5 panics per week across our userbase from dropping a `Dir`. It is possible that this issue is macOS specific (as all of our telemetry is from macOS today), but we are unfortunately not able to reproduce this panic internally.

The problem for us is that [this assertion](https://github.com/rust-lang/rust/blob/9c64068/library/std/src/sys/pal/unix/fs.rs#L874-L878) sometimes fails with `unexpected error during closedir: Os { code: 9, kind: Uncategorized, message: "Bad file descriptor" }`.

I expect that the standard library should not panic on situations that do happen in reality. In particular if you are trying to close a bad file descriptor, that seems "fine". For comparison a failure to close a file during the drop of a file descriptor is [silently ignored](https://github.com/rust-lang/rust/blob/9c64068/library/std/src/os/fd/owned.rs#L179) as I would expect. 

Not panicking is obviously not ideal; but in the context of a long-running interactive app like a text editor, we want to be able to take the trade off of higher uptime over lower visibility. 

For more discussion of possible solutions to fallible drops, #98338 has some good ideas, and I think if any are implemented we should do an equivalent thing for directories too (I particularly like the idea of a manual close that returns a result, and which causes the Drop to become a no-op). In the meantime, I think we should make rust programs crash less by ignoring this failure.

### Meta

`rustc --version --verbose`:
```
rustc 1.77.2 (25ef9e3d8 2024-04-09)
binary: rustc
commit-hash: 25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04
commit-date: 2024-04-09
host: aarch64-apple-darwin
release: 1.77.2
LLVM version: 17.0.6
```

(Though this was happening in earlier versions of rust too)

**Comment by workingjubilee** [comment]
@ConradIrwin Can you discuss the tracebacks for this, or does your telemetry lack sophisticated crash reporting?

**Comment by workingjubilee** [comment]
> I expect that the standard library should not panic on situations that do happen in reality. 

I know you meant something slightly aside from the reading of this, but I *do* expect the standard library to panic on indexing a slice with its length. So clearly the criteria has to be something more like
- whether it is reasonable to proceed anyways (arguable?)
- whether it is fixable (arguably not, if we get an error as joyously "descriptive" as EBADF).

**Comment by ConradIrwin** [comment]
Fair: I *do* expect panics on programmer error or in a situation where it's unsafe to continue. I am relatively (but it's obviously hard to be 100%) sure this is not programmer error, and I'm convinced that it'd be just fine if the program kept running (c.f. how this error is ignored when dropping a file descriptor).

The full back-trace is below. All line numbers in Zed code are relative to [v0.131.6](https://github.com/zed-industries/zed/tree/v0.131.6) and in rust [1.77.1](https://github.com/rust-lang/rust/tree/1.77.1). I'm not sure why the top line number is missed by [symbolicate](https://github.com/ConradIrwin/symbolicate), but the panic message narrows it down to [this assertion](https://github.com/rust-lang/rust/blob/9c64068/library/std/src/sys/pal/unix/fs.rs#L874-L878)

I believe the reason this panics at on line 4018 of worktree.rs in Zed's code is that  `child_paths` goes out of scope, which retains a `ReadDir` returned from `std::fs::read_dir()` (via smol) [here](https://github.com/zed-industries/zed/blob/cbeb120b624dcae040d0c96c2cb320ff0fd344b7/crates/fs/src/fs.rs#L343). I can't see any shenanigans on this codepath that would lead us to to invalidate the file descriptor, which leads me to believe that "this just happens sometimes" when interacting with the filesystem. (Though we are holding this across async/await, so it seems possible that something else is affecting the system's state, but it's very unclear as to what would cause this kind of a problem).

```
Panic `unexpected error during closedir: Os { code: 9, kind: Uncategorized, message: "Bad file descriptor" }` on thread 38 (com.apple.root.user-initiated-qos)
/rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/std/src/sys/pal/unix/fs.rs:0:	_$LT$std..sys..pal..unix..fs..Dir$u20$as$u20$core..ops..drop..Drop$GT$::drop::h84b244fb776f88f3
/rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/core/src/ptr/mod.rs:507:	core::ptr::drop_in_place::<std::path::PathBuf>
/rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/core/src/ptr/mod.rs:507:	core::ptr::drop_in_place::<std::sys::pal::unix::fs::InnerReadDir>
/rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/alloc/src/sync.rs:1756:	<alloc::sync::Arc<std::sys::pal::unix::fs::InnerReadDir>>::drop_slow
/rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/core/src/mem/mod.rs:394:	core::mem::size_of_val_raw::<dyn futures_core::stream::Stream<Item = core::result::Result<std::path::PathBuf, anyhow::Error>> + core::marker::Send>
/rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/core/src/alloc/layout.rs:201:	<core::alloc::layout::Layout>::for_value_raw::<dyn futures_core::stream::Stream<Item = core::result::Result<std::path::PathBuf, anyhow::Error>> + core::marker::Send>
/rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/alloc/src/boxed.rs:1241:	<alloc::boxed::Box<dyn futures_core::stream::Stream<Item = core::result::Result<std::path::PathBuf, anyhow::Error>> + core::marker::Send> as core::ops::drop::Drop>::drop
/rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/core/src/ptr/mod.rs:507:	core::ptr::drop_in_place::<alloc::boxed::Box<dyn futures_core::stream::Stream<Item = core::result::Result<std::path::PathBuf, anyhow::Error>> + core::marker::Send>>
/rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/core/src/ptr/mod.rs:507:	core::ptr::drop_in_place::<core::pin::Pin<alloc::boxed::Box<dyn futures_core::stream::Stream<Item = core::result::Result<std::path::PathBuf, anyhow::Error>> + core::marker::Send>>>
crates/worktree/src/worktree.rs:4018:	<worktree::BackgroundScanner>::scan_dir::{closure#0}
crates/worktree/src/worktree.rs:3754:	<worktree::BackgroundScanner>::scan_dirs::{closure#0}::{closure#0}::{closure#0}
crates/gpui/src/executor.rs:457:	<gpui::executor::Scope>::spawn::<<worktree::BackgroundScanner>::scan_dirs::{closure#0}::{closure#0}::{closure#0}>::{closure#0}
/rustc/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/core/src/task/poll.rs:51:	<core::task::poll::Poll<()>>::map::<core::result::Result<(), alloc::boxed::Box<dyn core::any::Any + core::marker::Send>>, core::result::Result<(), alloc::boxed::Box<dyn core::any::Any + core::marker::Send>>::Ok>
/Users/administrator/.cargo/registry/src/index.crates.io-6f17d22bba15001f/async-task-4.7.0/src/raw.rs:557:	<async_task::raw::RawTask<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = ()> + core::marker::Send>>, (), <gpui::executor::BackgroundExecutor>::spawn_internal<()>::{closure#0}, ()>>::run
libdispatch.dylib	_dispatch_client_callout
libdispatch.dylib	_dispatch_root_queue_drain
libdispatch.dylib	_dispatch_worker_thread2
libsystem_pthread.dylib	_pthread_wqthread
libsystem_pthread.dylib	start_wqthread
```

**Comment by workingjubilee** [comment]
Thank you for sharing!

I am not necessarily opposed to this change, I just felt it would be useful to know a bit more context.

Hmm. I wonder if, in another thread that is *not bound by the executor*, but does wait until *after* the worktree code has started up, destroying one of the directories in question would do it? That would make sense, and would explain why we get such a lovely "error" from the OS. That would be a classic case of "force majeure" that we typically do want to ignore, yes.

**Comment by the8472** [comment]
Caveat: `closedir` operates on a `DIR*`, not on a fd directly. So it's possible that the C library returns EBADF for something that's not an invalid file descriptor but some other invalid state in that wrapping struct.
Is the macos C library open source? Then we can check there. If not it'd require reproducing the issue and tracing syscalls to tell the difference.

**Comment by the8472** [comment]
Thanks for looking it up. I think that is enough evidence that that there's probably some library-UB going on in zed and we shouldn't try to silence it.

**Comment by joboet** [comment]
The only place in `zed` itself which calls `as_raw_fd` is [this code](https://github.com/zed-industries/zed/blob/1127b1a0deb2821c4cc688417ccb9348da65ef48/crates/terminal/src/pty_info.rs#L18-L65) for getting the PID of the child process running in the integrated terminal. While it's definitely very fishy (it's not tied to the lifetime of the PTY), it doesn't close anything so I doubt that this issue is caused there.

As for dependencies, it might help to get a complete backtrace to see which `ReadDir::drop` specifically is causing this.

**Comment by the8472** [comment]
It's not necessarily readdir that's causing this, it could be the victim of something else going rogue and closing the wrong fd numbers.

**Comment by joboet** [comment]
It must be... Though I'm hoping that there is enough locality between the `close` and the `read_dir` to point us in the direction.

**Comment by the8472** [comment]
I [have an idea](https://rust-lang.zulipchat.com/#narrow/stream/136281-t-opsem/topic/File.20descriptor.20randomization.20to.20catch.20some.20cases.20of.20IO.20U.2E.2E.2E/near/434205430) how to harden std types against these kinds of things, but it would be fairly expensive and imo only make sense for debug_assert builds. Randomization for file descriptor numbers. But I'm not sure if people would be on board with it.

**Comment by saethlin** [comment]
I think that is a very interesting idea and would support merging it as a debug assertion if it finds a bug in Zed.

**Comment by the8472** [comment]
Yeah, without rootcause or even reproducer I'm not claiming certainty. 

Another possibility that I see is some race in the `Arc` leading to a double-free.

**Comment by the8472** [comment]
@ConradIrwin I don't think this is a "this just happens sometimes" thing. File descriptors should not randomly vanish from a process. File descriptors are densely allocated resources holding critical data. If they were to get misnumbered under the nose of a process things could go seriously wrong, including data corruption.

So there could be a serious bug *somewhere* and that imo warrants further investigation rather than silencing it in the standard library. That it is some benging (although surprising) macos behavior is also _possibility_, but this should be proven before we silence it.

I have opened #124130 to abort on EBADF in OwnedFd. That should provide a signal to tell whether it's specific to ReadDir or file descriptors are getting clobbered in general. But you'll have to build Zed with a custom rustc build for that.

**Comment by workingjubilee** [comment]
> Arguably we should also not silently ignore EBADF for OwnedFd and only suppress other errors such as EIO.

Old MacDonald had a farm...

Yes, I'm in favor of us, at least for as long as we can tolerate it, silencing specifically whatever errors we *expect* to have happen and for us to be able to do nothing about, instead of "any error at all". I will simply trust @the8472 that EBADF is actually a highly specific error, in this context at least, and not merely the surface description of "something bad happened lol".

**Comment by ConradIrwin** [comment]
Thanks for all the input here, and particularly the reassurance that "this shouldn't happen".

It seems like the next steps are to try and narrow down:
* Is this a logic bug in Zed or it's dependencies.
* Is this a filesystem bug in macOS (or Fuse, or Samba - which I know we have some users using).

For the in-process case it seems like a "bad file descriptor" means "one that the OS doesn't think you have", so either:
* Something is generating arbitrary numbers and sneaking them into this Dir object (seems very unlikely)
* Something is closing arbitrary numbers as file descriptors and sometimes this happens to close an open Dir object (seems very unlikely).
* Something is closing a file-descriptor twice, and between the first and second close the file-descriptor is re-used by the Dir open, so by the time we come to drop the Dir we have lost it (seems most plausible).
* (maybe) Something is closing all file-descriptors on app shutdown, but rust code continues to run after that?
* ...your idea here...

Although most of our code is rust, we do like to the Cocoa runtime, and a few other non-rust libraries, so the search space is rather large. Ideas for narrowing it down would be very appreciated! (In rust, I think we just need to audit all our crates for `AsRawFd/IntoRawFd`, and assuming no rust bugs, everything else should be "safe" according to https://rust-lang.github.io/rfcs/3128-io-safety.html?)

For the filesystem bug case, I haven't yet tried reproducing this on other filesystems. I know we have users using Samba at least, and Fuse/sshfs seems quite probable too, so that might be a fruitful path.

If anyone's interested in this kind of spelunking, I'd love a second brain while working through this: https://calendly.com/conradirwin/pairing, otherwise any interesting hypothesis that we can test for how to get a reproduction would be awesome.

If this does turn out to be a bug on the file system side, then I think we do need to *not* panic in rust. If it turns out to be a bug on the user-space side, then we'll be glad to have tracked it down, and the current behavior seems ok (until we do find a filesystem bug, at which point we'll want to be back to square one).

In parallel we might want to explore ideas from #98338 on how we can handle fallible drops for things like this.

**Comment by the8472** [comment]
> (In rust, I think we just need to audit all our crates for AsRawFd/IntoRawFd, and assuming no rust bugs, everything else should be "safe" according to https://rust-lang.github.io/rfcs/3128-io-safety.html?)

Yes, that's the intent.

I'm currently writing a FUSE testcase on linux to see if I can get a EBADF propagated from one of its replies to a `close()`.

**Comment by ConradIrwin** [comment]
Thanks! I should be able to do something similar for macFUSE / Samba next week.

**Comment by the8472** [comment]
```
openat(AT_FDCWD, "./fuse-test/foo", O_RDONLY|O_CLOEXEC) = 5
close(5)                                = -1 EBADF (Bad file descriptor)
```

Hope has been canceled once again.

![hope out of service](https://github.com/rust-lang/rust/assets/1065730/09f49af0-2b44-4d9d-a16d-5f6eca86c8f7)

Test program

```rust
use std::{fs::File, thread, time::{Duration, UNIX_EPOCH}};
use fuser::{FileAttr, FileType};

struct TestFs {}

const DUMMY_ATTR: FileAttr = FileAttr {
    ino: 2,
    size: 13,
    blocks: 1,
    atime: UNIX_EPOCH,
    mtime: UNIX_EPOCH,
    ctime: UNIX_EPOCH,
    crtime: UNIX_EPOCH,
    kind: FileType::RegularFile,
    perm: 0o644,
    nlink: 1,
    uid: 501,
    gid: 20,
    rdev: 0,
    flags: 0,
    blksize: 512,
};

impl fuser::Filesystem for TestFs {
    fn init(&mut self, _req: &fuser::Request<'_>, _config: &mut fuser::KernelConfig) -> Result<(), libc::c_int> {
        Ok(())
    }

    fn lookup(&mut self, _req: &fuser::Request<'_>, parent: u64, name: &std::ffi::OsStr, reply: fuser::ReplyEntry) {
        let d = Duration::from_secs(1);
        reply.entry(&d, &DUMMY_ATTR, 0);
    }

    fn open(&mut self, _req: &fuser::Request<'_>, _ino: u64, _flags: i32, reply: fuser::ReplyOpen) {
        reply.opened(1, 0);
    }

    fn flush(&mut self, _req: &fuser::Request<'_>, ino: u64, fh: u64, lock_owner: u64, reply: fuser::ReplyEmpty) {
        reply.error(libc::EBADF);
    }
}

fn main() {
    let fs = TestFs {};

    let t = thread::spawn(|| {
        //let options = [MountOption::RO, MountOption::AutoUnmount, MountOption::AllowRoot];
        fuser::mount2(fs, "mnt-test", &[]).expect("mount failed");
    });

    thread::sleep(Duration::from_millis(500));

    let f = File::open("./mnt-test/foo").unwrap();
    drop(f);

    t.join().unwrap();
}
```

**Comment by the8472** [comment]
If this replicates on macos and users have some janky fuse drivers that'd be a likely suspect. Though those drivers should have a bug filed against them.

**Comment by ConradIrwin** [comment]
A quick (though inconclusive) update: Although I'm not certain it doesn't reproduce, I was unable to reproduce with the example program (it fails before it gets to the end), and my testing with Zed itself, I could only reproduce https://github.com/libgit2/libgit2/pull/6796, not this error with `sshfs`.

Still to try: updating the example program to run to completion, and setting up a Samba server to test that one.

**Comment by bjorn3** [comment]
I just tried compiling Zed on Linux with the rustc_codegen_cranelift backend. It crashes reliably with:

```
fatal runtime error: IO Safety violation: owned file descriptor already closed

Thread 1 "Zed" received signal SIGABRT, Aborted.
__pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at ./nptl/pthread_kill.c:44
44      ./nptl/pthread_kill.c: Bestand of map bestaat niet.
(gdb) bt
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at ./nptl/pthread_kill.c:44
#1  0x00007ffff786fe8f in __pthread_kill_internal (signo=6, threadid=<optimized out>) at ./nptl/pthread_kill.c:78
#2  0x00007ffff7820fb2 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007ffff780b472 in __GI_abort () at ./stdlib/abort.c:79
#4  0x0000555562b7e5d1 in std::sys::pal::unix::abort_internal () at library/std/src/sys/pal/unix/mod.rs:366
#5  0x0000555560e37664 in std::sys::pal::unix::fs::debug_assert_fd_is_open () at /home/bjorn/Projects/cg_clif2/build/stdlib/library/std/src/rt.rs:43
#6  0x0000555560dffad6 in std::os::fd::owned::{impl#7}::drop () at /home/bjorn/Projects/cg_clif2/build/stdlib/library/std/src/os/fd/owned.rs:181
#7  0x0000555560e38069 in core::ptr::drop_in_place<std::os::fd::owned::OwnedFd> () at /home/bjorn/Projects/cg_clif2/build/stdlib/library/core/src/ptr/mod.rs:542
#8  0x000055555b0f8dde in core::ptr::drop_in_place<std::sys::pal::unix::fd::FileDesc> () at /home/bjorn/Projects/cg_clif2/build/stdlib/library/core/src/ptr/mod.rs:542
#9  0x000055555b0f928c in core::ptr::drop_in_place<std::sys::pal::unix::process::process_common::Stdio> () at /home/bjorn/Projects/cg_clif2/build/stdlib/library/core/src/ptr/mod.rs:542
#10 0x000055555b0f8856 in core::ptr::drop_in_place<core::option::Option<std::sys::pal::unix::process::process_common::Stdio>> ()
    at /home/bjorn/Projects/cg_clif2/build/stdlib/library/core/src/ptr/mod.rs:542
#11 0x000055555b0f930f in core::ptr::drop_in_place<std::sys::pal::unix::process::process_common::Command> () at /home/bjorn/Projects/cg_clif2/build/stdlib/library/core/src/ptr/mod.rs:542
#12 0x000055555b0f8d67 in core::ptr::drop_in_place<std::process::Command> () at /home/bjorn/Projects/cg_clif2/build/stdlib/library/core/src/ptr/mod.rs:542
#13 0x000055555b10be12 in alacritty_terminal::tty::unix::new () at /home/bjorn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/alacritty_terminal-0.23.0/src/tty/unix.rs:285
#14 0x000055555afdc160 in terminal::{impl#7}::new () at crates/terminal/src/terminal.rs:377
#15 0x000055555a49c2fb in project::terminals::{impl#0}::create_terminal () at crates/project/src/terminals.rs:167
```

```
$ dist/rustc-clif -vV
rustc 1.80.0-nightly (b1ec1bd65 2024-05-18)
binary: rustc
commit-hash: b1ec1bd65f89c1375d2cf2fb733a87ef390276d3
commit-date: 2024-05-18
host: x86_64-unknown-linux-gnu
release: 1.80.0-nightly
Cranelift version: 0.108.0
```

Upon further investigation, I found https://github.com/alacritty/alacritty/pull/7996, which fixes the crash I found here. I think it is very likely that the bug this fixes is responsible for the panic described in this issue too.

**Comment by ConradIrwin** [comment]
If you can reproduce this reliably, I’d love to debug it together - do you have time to talk the first week of July? If so please reach out to conrad@zed.dev

In my mind there’s quite a chance this is a bug in zed or its dependencies, but it’s also possible that this is a failure introduced by the filesystem.

things I’d like to try:
* adding logging to open/close fds to see if we can find a double close
* Knowing more about the file system you’re using

**Comment by ConradIrwin** [comment]
Nvm, I just read the attached issue. I’ll pull that into zed and see if it fixes out problems

**Comment by bjorn3** [comment]
I can confirm that using the current master branch of alacritty-terminal fixes the IO safety crash with rustc_codegen_cranelift. It does still crash, but that is because of some unimplemented things in rustc_codegen_cranelift (https://github.com/rust-lang/rustc_codegen_cranelift/issues/1492), not a bug in Zed or any of it's dependencies.

```diff
diff --git a/crates/terminal/Cargo.toml b/crates/terminal/Cargo.toml
index ef6260730..3ce869597 100644
--- a/crates/terminal/Cargo.toml
+++ b/crates/terminal/Cargo.toml
@@ -14,7 +14,7 @@ doctest = false
 
 
 [dependencies]
-alacritty_terminal = "0.23"
+alacritty_terminal = { git = "https://github.com/alacritty/alacritty", version = "0.24.1-dev" }
 anyhow.workspace = true
 collections.workspace = true
 dirs = "4.0.0"
```

https://github.com/alacritty/alacritty/pull/7996 hasn't been published to crates.io yet, so I had to use a git dependency.

**Comment by ConradIrwin** [comment]
We've been running the fixed alacritty in preview for a week, and on stable for a few days, and have seen 0 instances of this - this either means we've been unusually lucky, or this is now solved.

Thank you very much to @the8472 for adding the debug asserts, and to @jakobhellermann for figuring out the problem.

**Comment by daandemeyer** [comment]
I ran into this on notify's CI, except it showed up as an abort rather than the closedir panic:

```
test poll::tests::remove_file ... ok
fatal runtime error: IO Safety violation: owned file descriptor already closed, aborting
error: test failed, to rerun pass `-p notify --lib`

Caused by:
  process didn't exit successfully: `.../deps/notify-f7d57847dd78bfa8` (signal: 6, SIGABRT: process abort signal)
```

Full job: https://github.com/notify-rs/notify/actions/runs/30158197016/job/89679165708

I'm fairly sure it's the same underlying problem. `debug_assert_fd_is_open` in `library/std/src/sys/fs/unix.rs` does an `fcntl(fd, F_GETFD)` and aborts on EBADF. It has two callers, `OwnedFd::drop` and `DirStream::drop`, and in `DirStream::drop` it runs immediately before the `closedir` this issue is about:

```rust
let fd = unsafe { libc::dirfd(self.0) };
debug_assert_fd_is_open(fd);

let r = unsafe { libc::closedir(self.0) };
assert!(
    r == 0 || crate::io::Error::last_os_error().is_interrupted(),
    "unexpected error during closedir: {:?}",
    ...
);
```

Environment:

```
rustc 1.97.1 (8bab26f4f 2026-07-14) stable
aarch64-apple-darwin
GitHub Actions macos-26-arm64
cargo test --workspace --exclude examples --no-default-features --features macos_fsevent --features tokio
debug profile
```

The reproducer isn't deterministic, but it is narrow. The test that provokes it does three things:

1. creates an FSEvents watcher (`FSEventStreamCreate` plus `FSEventStreamScheduleWithRunLoop` on a dedicated run loop thread)
2. creates 4097 paths of the form `dir_N/subdir`, so 8194 directories, and asks FSEvents to watch all 4097 at once, which fails because a stream seems to cap out around 4096 paths (https://github.com/fsnotify/fsevents/issues/48)
3. deletes that whole tree with `remove_dir_all` while the stream is still running

The abort is rarer than the panic. I found one occurrence across the CI runs I looked at, whereas the panic recurred often enough that someone added the workaround.

Two things we ruled out on our side:

notify doesn't close any raw file descriptors on this platform. The inotify and kqueue backends are the only ones touching `RawFd`/`OwnedFd` and neither is compiled in the macOS FSEvents configuration. The FSEvents backend only holds CoreFoundation objects. The descriptors notify owns here all come from plain `std::fs::File`.

`remove_dir_all`'s own ownership looks right to me. `fdreaddir` passes the fd to `fdopendir` and then calls `into_raw_fd` on the `OwnedFd`, so `closedir` becomes the only closer, and `remove_dir_all_recursive` just borrows that raw fd for `openat` and `unlinkat`. The `chmod_harder` code that does hold an `OwnedFd` is cfg'd to Linux and Android.

That leaves something else in the process closing the descriptor. Both reports I know of, this one and the Zed report that opened the issue, are macOS applications using FSEvents.

So, two questions:

1. Is a stray `close()` coming from a system framework a known failure mode here? Is there a supported way to instrument it, something like interposing on `close` or a descriptor guard equivalent to Guard Malloc?
2. Would a standalone reproducer with no notify dependency be useful, just an FSEventStream over a large tree plus a concurrent `remove_dir_all`? I can write that if it would help, since I gather this hasn't been reproducible internally so far.
`

**Comment by the8472** [comment]
> Is a stray close() coming from a system framework a known failure mode here?

We have had no such reports.

If there are things that close FDs that they don't own it does not necessarily trigger the abort anywhere near the code or on the file descriptor that's responsible because when the numbers start getting reused it can become hopelessly confused which object owns which FD, especially in concurrent programs.
If you get into this situation you may already be facing data corruption from stray writes into the wrong file descriptors. It is irrecoverable global state corruption, which is why we abort the process.

>  Is there a supported way to instrument it, something like interposing on close or a descriptor guard equivalent to Guard Malloc?

A macos equivalent of `strace` is what I'd reach for to find double-closes.


> Would a standalone reproducer with no notify dependency be useful, just an FSEventStream over a large tree plus a concurrent remove_dir_all? I can write that if it would help, since I gather this hasn't been reproducible internally so far.

Wrong issue? This is the first time that FSEvents has been brought up here. If you think there's an issue in Rust or want input from us please open a new one.
But sure, a minimal reproducer will help anyone looking into it.

**Comment by daandemeyer** [comment]
This is a bug in fsevents, has nothing to do with rust, apologies for the noise

**Comment by daandemeyer** [comment]
I tracked down the bug and posted a reproducer: https://github.com/daandemeyer/fsevents-fd-repro, nothing to do with rust, but posting it here in case someone else finds this thread while looking into similar issues.

**Comment by RalfJung** [comment]
That's at least the 2nd time that this assertion found an actual bug :)

**Comment by the8472** [comment]
Incomplete list of trophies from searching public github issues:

- https://github.com/alacritty/alacritty/pull/7996 (this issue that started it) 
- https://github.com/stack-rs/rattan/issues/80
- https://github.com/kata-containers/kata-containers/issues/11054
- https://github.com/vorner/signal-hook/issues/200
- https://github.com/esa-tu-darmstadt/tapasco/pull/404
- https://github.com/stratis-storage/stratisd/issues/3946
- https://github.com/cloud-hypervisor/cloud-hypervisor/issues/6485
